### PR TITLE
tidb: eliminate runtime.morestack from getTimestamp call stack

### PR DESCRIPTION
**Pull request by tiancaiamao**
By reserved 4KB memory on the stack, we can avoid runtime.morestack when getTimestamp in `store.Begin`.

Test on local machine with 1 pd and 1 tikv (not under stress testing),
Eliminate the runtime.morestack make `store.Begin` response time from 157000ns to 152000ns.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @siddontang @coocood

File: session.go [hunk 1]
```diff
@@ -991,27 +995,57 @@ func (s *session) loadCommonGlobalVariablesIfNeeded() error {
...
+	future.Add(1)
+	s.txnCh <- future
+	return future
+}
+
+//go:noinline
+func asyncGetTSWorker(store kv.Storage, ch chan *txnFuture) {
+	// Reserved 4KB memory on the stack to avoid runtime.morestack
+	var buf [4 << 10]byte
+	if false {
```
Comment (coocood) [inline session.go]: This may be optimized by future Go compiler.

File: session.go [hunk 2]
```diff
@@ -106,7 +106,8 @@ type session struct {
 	// It's used by ShowProcess(), and should be modified atomically.
 	processInfo atomic.Value
 	txn         kv.Transaction // current transaction
-	txnCh       chan *txnWithErr
+	txnFuture   *txnFuture
+	txnCh       chan *txnFuture
```
Comment (coocood) [inline session.go]: s/txnCh/txnFutureCh

File: session.go [hunk 3]
```diff
@@ -991,27 +995,57 @@ func (s *session) loadCommonGlobalVariablesIfNeeded() error {
 	return nil
 }
 
-type txnWithErr struct {
+// txnFuture is a promise, which promises to return a txn in future.
+type txnFuture struct {
```
Comment (coocood) [inline session.go]: Since you defined the type, why not define a method wait and get the transaction?

Comment (tiancaiamao) [inline session.go]:
```
s.txn != nil
s.txn.Valid()
s.txnFuture
```

It's quite boring, we can just provide txnFuture and provide method on it...
but maybe we can leave it to later refactor, not in this PR

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991  @zimulala @XuHuaiyu   @shenli

File: statistics/update_test.go [hunk 4]
```diff
@@ -148,18 +148,18 @@ func (s *testStatsUpdateSuite) TestMultiSession(c *C) {
 		testKit2.MustExec("delete from test.t1 limit 1")
 	}
 
-	err = testKit.Se.Close()
```
Comment (hanfei1991) [inline statistics/update_test.go]: Don't move this code. This is to test the session will update the stats even if it is closed.

Comment (tiancaiamao) [inline statistics/update_test.go]: Session should not be used anymore, if it's Closed !!!
Close means all resource bind to it should be released, if you violate the constraint, that's really bad design.

**Pull request merged**